File: protocols/proven-kms/ffi/zig/src/kms.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// kms.zig — Zig FFI implementation of proven-kms.
//
// Implements the Key Management Server primitive with:
//   - Slot-based key context management (up to 64 concurrent keys)
//   - KMIP-style key lifecycle state machine (PreActive -> Active ->
//     Deactivated -> Destroyed, with Compromised branches)
//   - Operation tracking (create, encrypt, decrypt, sign, verify, etc.)
//   - Algorithm and object type metadata
//   - Thread-safe via mutex on global state
//
// Tag values MUST match:
//   - Idris2 ABI (src/KMSABI/Layout.idr)
//   - C header   (generated/abi/kms.h)

const std = @import("std");

// ── Enums (matching Idris2 Layout.idr tag assignments exactly) ──────────

/// ObjectType — matches objectTypeToTag
pub const ObjectType = enum(u8) {
    symmetric_key = 0,
    public_key = 1,
    private_key = 2,
    secret_data = 3,
    certificate = 4,
    opaque_data = 5,
};

/// Operation — matches operationToTag
pub const Operation = enum(u8) {
    create = 0,
    get = 1,
    activate = 2,
    revoke = 3,
    destroy = 4,
    locate = 5,
    register = 6,
    rekey = 7,
    encrypt = 8,
    decrypt = 9,
    sign = 10,
    verify = 11,
    wrap = 12,
    unwrap = 13,
    mac = 14,
};

/// KeyState — matches keyStateToTag
pub const KeyState = enum(u8) {
    pre_active = 0,
    active = 1,
    deactivated = 2,
    compromised = 3,
    destroyed = 4,
    destroyed_compromised = 5,
};

/// Algorithm — matches algorithmToTag
pub const Algorithm = enum(u8) {
    aes128 = 0,
    aes256 = 1,
    rsa2048 = 2,
    rsa4096 = 3,
    ecdsa_p256 = 4,
    ecdsa_p384 = 5,
    ed25519 = 6,
    chacha20_poly1305 = 7,
    hmac_sha256 = 8,
};

/// KMSError — matches kmsErrorToTag
pub const KMSError = enum(u8) {
    ok = 0,
    invalid_slot = 1,
    not_active = 2,
    invalid_transition = 3,
    operation_denied = 4,
    capacity_exhausted = 5,
    unsupported_alg = 6,
    key_destroyed = 7,
};

// ── Key Context instance ────────────────────────────────────────────────

const KeyCtx = struct {
    /// Whether this slot is in use.
    active: bool,
    /// Current lifecycle state.
    state: KeyState,
    /// Type of cryptographic object stored in this slot.
    obj_type: ObjectType,
    /// Cryptographic algorithm associated with the key.
    algorithm: Algorithm,
    /// Last error code (255 = no error).
    last_error: u8,
    /// Number of operations performed on this key.
    operation_count: u32,
};

// ── Global state (slot-based, mutex-protected) ──────────────────────────

const MAX_CONTEXTS: usize = 64;

const empty_ctx: KeyCtx = .{
    .active = false,
    .state = .pre_active,
    .obj_type = .symmetric_key,
    .algorithm = .aes128,
    .last_error = 255,
    .operation_count = 0,
};

var contexts: [MAX_CONTEXTS]KeyCtx = [_]KeyCtx{empty_ctx} ** MAX_CONTEXTS;
var mutex: std.Thread.Mutex = .{};

// ── Helpers ─────────────────────────────────────────────────────────────

/// Validate and return a pointer to an active context, or null.
fn getActive(slot: c_int) ?*KeyCtx {
    if (slot < 0 or slot >= MAX_CONTEXTS) return null;
    const idx: usize = @intCast(slot);
    if (!contexts[idx].active) return null;
    return &contexts[idx];
}

// ── ABI version ─────────────────────────────────────────────────────────

/// ABI version — must match KMSABI.Foreign.abiVersion (currently 1).
pub export fn kms_abi_version() callconv(.c) u32 {
    return 1;
}

// ── Lifecycle ───────────────────────────────────────────────────────────

/// Create a new key context.
/// Returns slot index (0-63) or -1 if no slots available or invalid params.
pub export fn kms_create(obj_type: u8, algorithm: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    // Validate obj_type (0-5)
    if (obj_type > 5) return -1;
    // Validate algorithm (0-8)
    if (algorithm > 8) return -1;

    for (&contexts, 0..) |*ctx, i| {
        if (!ctx.active) {
            ctx.* = empty_ctx;
            ctx.active = true;
            ctx.obj_type = @enumFromInt(obj_type);
            ctx.algorithm = @enumFromInt(algorithm);
            return @intCast(i);
        }
    }
    return -1; // all slots occupied
}

/// Destroy a key context, freeing its slot.
/// Safe to call with any slot index (invalid slots are no-ops).
pub export fn kms_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();

    if (slot < 0 or slot >= MAX_CONTEXTS) return;
    const idx: usize = @intCast(slot);
    contexts[idx].active = false;
}

// ── State queries ───────────────────────────────────────────────────────

/// Get the current KeyState tag for a slot.
/// Returns PreActive (0) for invalid/inactive slots.
pub export fn kms_get_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.state);
}

/// Get the ObjectType tag for a slot.
/// Returns SymmetricKey (0) for invalid/inactive slots.
pub export fn kms_get_object_type(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.obj_type);
}

/// Get the Algorithm tag for a slot.
/// Returns AES128 (0) for invalid/inactive slots.
pub export fn kms_get_algorithm(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.algorithm);
}

/// Get the number of operations performed on this key.
pub export fn kms_get_operation_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return ctx.operation_count;
}

/// Get the last KMSError tag, or 255 if no error.
pub export fn kms_get_last_error(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 255;
    return ctx.last_error;
}

// ── State transitions ───────────────────────────────────────────────────

/// Advance a key to a new state, validating the transition.
/// Returns KMSError tag.
pub export fn kms_transition(slot: c_int, new_state: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(KMSError.invalid_slot);

    const from = @intFromEnum(ctx.state);
    if (kms_can_transition(from, new_state) == 0) {
        ctx.last_error = @intFromEnum(KMSError.invalid_transition);
        return @intFromEnum(KMSError.invalid_transition);
    }

    ctx.state = @enumFromInt(new_state);
    ctx.last_error = 255;
    return @intFromEnum(KMSError.ok);
}

/// Record an operation on the key.
/// Key must be in Active state for cryptographic operations.
/// Create (0), Get (1), Locate (5), Register (6) are allowed in any non-destroyed state.
/// Returns KMSError tag.
pub export fn kms_perform_operation(slot: c_int, operation: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(KMSError.invalid_slot);

    // Validate operation (0-14)
    if (operation > 14) {
        ctx.last_error = @intFromEnum(KMSError.invalid_transition);
        return @intFromEnum(KMSError.invalid_transition);
    }

    // Destroyed keys cannot have any operations performed
    if (ctx.state == .destroyed or ctx.state == .destroyed_compromised) {
        ctx.last_error = @intFromEnum(KMSError.key_destroyed);
        return @intFromEnum(KMSError.key_destroyed);
    }

    // Cryptographic operations (encrypt, decrypt, sign, verify, wrap,
    // unwrap, mac) require Active state
    if (operation >= 8 and operation <= 14) {
        if (ctx.state != .active) {
            ctx.last_error = @intFromEnum(KMSError.operation_denied);
            return @intFromEnum(KMSError.operation_denied);
        }
    }

    ctx.operation_count += 1;
    ctx.last_error = 255;
    return @intFromEnum(KMSError.ok);
}

// ── Stateless transition validation ─────────────────────────────────────

/// Check whether a key state transition is valid.
/// Returns 1 if valid, 0 if not.
/// Models KMIP key lifecycle per SP 800-57.
pub export fn kms_can_transition(from: u8, to: u8) callconv(.c) u8 {
    // PreActive (0) -> Active (1)
    if (from == 0 and to == 1) return 1;
    // PreActive (0) -> Destroyed (4) (never activated, discard)
    if (from == 0 and to == 4) return 1;
    // Active (1) -> Deactivated (2)
    if (from == 1 and to == 2) return 1;
    // Active (1) -> Compromised (3)
    if (from == 1 and to == 3) return 1;
    // Active (1) -> Destroyed (4)
    if (from == 1 and to == 4) return 1;
    // Deactivated (2) -> Destroyed (4)
    if (from == 2 and to == 4) return 1;
    // Deactivated (2) -> Compromised (3)
    if (from == 2 and to == 3) return 1;
    // Compromised (3) -> DestroyedCompromised (5)
    if (from == 3 and to == 5) return 1;
    // PreActive (0) -> Compromised (3) (key compromised before activation)
    if (from == 0 and to == 3) return 1;
    return 0;
}

File: bindings/rust/src/kms.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! Key Management Service types for the proven-servers ABI.
//!
//! Formally verified KMS types (KMIP-compatible).
//! Mirrors the Idris2 module `KmsABI.Types`.
//!
//! - `ObjectType` -- Managed cryptographic object types.
//! - `Operation` -- KMS operations.
//! - `KeyState` -- Key lifecycle states (KMIP).
//! - `KmsAlgorithm` -- Cryptographic algorithms.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// Key Management Service Constants
// ===========================================================================

/// Standard KMIP port.
pub const KMS_PORT: u16 = 5696;

// ===========================================================================
// ObjectType (tags 0-5)
// ===========================================================================

/// Managed cryptographic object types.
///
/// Matches `ObjectType` in `KmsABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ObjectType {
    /// SymmetricKey (tag 0).
    SymmetricKey = 0,
    /// PublicKey (tag 1).
    PublicKey = 1,
    /// PrivateKey (tag 2).
    PrivateKey = 2,
    /// SecretData (tag 3).
    SecretData = 3,
    /// Certificate (tag 4).
    Certificate = 4,
    /// OpaqueData (tag 5).
    OpaqueData = 5,
}

impl ObjectType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::SymmetricKey),
            1 => Some(Self::PublicKey),
            2 => Some(Self::PrivateKey),
            3 => Some(Self::SecretData),
            4 => Some(Self::Certificate),
            5 => Some(Self::OpaqueData),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ObjectType; 6] = [
        Self::SymmetricKey, Self::PublicKey, Self::PrivateKey, Self::SecretData, Self::Certificate, Self::OpaqueData,
    ];
}

impl fmt::Display for ObjectType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Operation (tags 0-14)
// ===========================================================================

/// KMS operations.
///
/// Matches `Operation` in `KmsABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Operation {
    /// Create (tag 0).
    Create = 0,
    /// Get (tag 1).
    Get = 1,
    /// Activate (tag 2).
    Activate = 2,
    /// Revoke (tag 3).
    Revoke = 3,
    /// Destroy (tag 4).
    Destroy = 4,
    /// Locate (tag 5).
    Locate = 5,
    /// Register (tag 6).
    Register = 6,
    /// Rekey (tag 7).
    Rekey = 7,
    /// Encrypt (tag 8).
    Encrypt = 8,
    /// Decrypt (tag 9).
    Decrypt = 9,
    /// Sign (tag 10).
    Sign = 10,
    /// Verify (tag 11).
    Verify = 11,
    /// Wrap (tag 12).
    Wrap = 12,
    /// Unwrap (tag 13).
    Unwrap = 13,
    /// MAC (tag 14).
    Mac = 14,
}

impl Operation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Create),
            1 => Some(Self::Get),
            2 => Some(Self::Activate),
            3 => Some(Self::Revoke),
            4 => Some(Self::Destroy),
            5 => Some(Self::Locate),
            6 => Some(Self::Register),
            7 => Some(Self::Rekey),
            8 => Some(Self::Encrypt),
            9 => Some(Self::Decrypt),
            10 => Some(Self::Sign),
            11 => Some(Self::Verify),
            12 => Some(Self::Wrap),
            13 => Some(Self::Unwrap),
            14 => Some(Self::Mac),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// Whether this is a cryptographic operation.
    pub fn is_crypto_op(self) -> bool {
        matches!(self, Self::Encrypt | Self::Decrypt | Self::Sign | Self::Verify | Self::Wrap | Self::Unwrap | Self::Mac)
    }

    /// Whether this is a key lifecycle operation.
    pub fn is_lifecycle_op(self) -> bool {
        matches!(self, Self::Create | Self::Activate | Self::Revoke | Self::Destroy | Self::Rekey)
    }

    /// All variants of this type.
    pub const ALL: [Operation; 15] = [
        Self::Create, Self::Get, Self::Activate, Self::Revoke, Self::Destroy, Self::Locate, Self::Register, Self::Rekey, Self::Encrypt, Self::Decrypt, Self::Sign, Self::Verify, Self::Wrap, Self::Unwrap, Self::Mac,
    ];
}

impl fmt::Display for Operation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// KeyState (tags 0-5)
// ===========================================================================

/// Key lifecycle states (KMIP).
///
/// Matches `KeyState` in `KmsABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum KeyState {
    /// PreActive (tag 0).
    PreActive = 0,
    /// Active (tag 1).
    Active = 1,
    /// Deactivated (tag 2).
    Deactivated = 2,
    /// Compromised (tag 3).
    Compromised = 3,
    /// Destroyed (tag 4).
    Destroyed = 4,
    /// DestroyedCompromised (tag 5).
    DestroyedCompromised = 5,
}

impl KeyState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::PreActive),
            1 => Some(Self::Active),
            2 => Some(Self::Deactivated),
            3 => Some(Self::Compromised),
            4 => Some(Self::Destroyed),
            5 => Some(Self::DestroyedCompromised),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// Whether the key can be used for cryptographic operations.
    pub fn is_usable(self) -> bool {
        matches!(self, Self::Active)
    }

    /// All variants of this type.
    pub const ALL: [KeyState; 6] = [
        Self::PreActive, Self::Active, Self::Deactivated, Self::Compromised, Self::Destroyed, Self::DestroyedCompromised,
    ];
}

impl fmt::Display for KeyState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// KmsAlgorithm (tags 0-8)
// ===========================================================================

/// Cryptographic algorithms.
///
/// Matches `KmsAlgorithm` in `KmsABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum KmsAlgorithm {
    /// AES-128 (tag 0).
    Aes128 = 0,
    /// AES-256 (tag 1).
    Aes256 = 1,
    /// RSA-2048 (tag 2).
    Rsa2048 = 2,
    /// RSA-4096 (tag 3).
    Rsa4096 = 3,
    /// ECDSA P-256 (tag 4).
    EcdsaP256 = 4,
    /// ECDSA P-384 (tag 5).
    EcdsaP384 = 5,
    /// Ed25519 (tag 6).
    Ed25519 = 6,
    /// Chacha20Poly1305 (tag 7).
    Chacha20Poly1305 = 7,
    /// HMAC-SHA256 (tag 8).
    HmacSha256 = 8,
}

impl KmsAlgorithm {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Aes128),
            1 => Some(Self::Aes256),
            2 => Some(Self::Rsa2048),
            3 => Some(Self::Rsa4096),
            4 => Some(Self::EcdsaP256),
            5 => Some(Self::EcdsaP384),
            6 => Some(Self::Ed25519),
            7 => Some(Self::Chacha20Poly1305),
            8 => Some(Self::HmacSha256),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [KmsAlgorithm; 9] = [
        Self::Aes128, Self::Aes256, Self::Rsa2048, Self::Rsa4096, Self::EcdsaP256, Self::EcdsaP384, Self::Ed25519, Self::Chacha20Poly1305, Self::HmacSha256,
    ];
}

impl fmt::Display for KmsAlgorithm {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn object_type_roundtrip() {
        for v in ObjectType::ALL {
            let tag = v.to_tag();
            let decoded = ObjectType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ObjectType::from_tag(6).is_none());
    }

    #[test]
    fn operation_roundtrip() {
        for v in Operation::ALL {
            let tag = v.to_tag();
            let decoded = Operation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Operation::from_tag(15).is_none());
    }

    #[test]
    fn key_state_roundtrip() {
        for v in KeyState::ALL {
            let tag = v.to_tag();
            let decoded = KeyState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(KeyState::from_tag(6).is_none());
    }

    #[test]
    fn kms_algorithm_roundtrip() {
        for v in KmsAlgorithm::ALL {
            let tag = v.to_tag();
            let decoded = KmsAlgorithm::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(KmsAlgorithm::from_tag(9).is_none());
    }

    #[test]
    fn constants_match_idris() {
        assert_eq!(KMS_PORT, 5696);
    }

}
